compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala

```diff
@@ -54,10 +55,26 @@ class PCPCheckAndHeal(@constructorOnly ictx: Context) extends TreeMapWithStages(
 
   /** Transform quoted trees while maintaining phase correctness */
   override protected def transformQuotation(body: Tree, quote: Tree)(implicit ctx: Context): Tree = {
+    val taggedTypes = new PCPCheckAndHeal.QuoteTypeTags(quote.span)(using ctx)
+    given Context =
+      if level(ctx) == 0 then contextWithQuoteTypeTags(taggedTypes)(ctx)
+      else ctx
+
     if (ctx.property(InAnnotation).isDefined)
       ctx.error("Cannot have a quote in an annotation", quote.sourcePos)
-    val body1 = transform(body)(quoteContext)
-    super.transformQuotation(body1, quote)
+
+    val body1 =
+      transform(body)(quoteContext)
+
+    val body2 =
+      if level != 0 then body1
```

Comment:
Minor: it seems to me the logic will be more clear if we handle `level == 0` in one branch, and `level != 0` in another at the top-level.